Step3 pull request 요청드립니다.

hyunhyun opened the pull request:
안녕하세요 step3-로또(자동) 리뷰 요청드립니다. 
구현하는데 시간이 꽤 걸렸네요
리뷰 잘 부탁드리겠습니다~ 감사합니다!🙂

hyunhyun committed:
refactor(StringAddCalculator) :

 - string -> pattern 상수 변경
 - 메소드명 변경(numberValidate -> validateNumber)
 - 상수명 변경
 - 메소드 throws 삭제

hyunhyun committed:
docs(README.md) : step3 기능요구사항 정리

hyunhyun committed:
feat : 로또 한장 생성 구현, 테스트코드 작성

hyunhyun committed:
feat : 구입금액에 따른 로또 여러건 생성

- LottoTest : 금액 입력 로또 개수 확인 테스트코드
- LottoMachine : 로또 여러건 생성, 구매내역 출력
- Lottos : 로또 여러장 저장 컬렉션
- Lotto : 로또 내용 출력

hyunhyun committed:
refactor : 로또 구매 정보 출력 클래스 분리

hyunhyun committed:
feat : 구입금액 입력, 유효성 검사 추가

hyunhyun committed:
feat : 당첨 개수 구하기, 총 당첨금액 구하기, 당첨결과 출력

Lotto : 당첨로또 비교 당첨개수 구하기
Lottos : 총 당첨금액 구하기, 등수별 개수 구하기
ResultView : 등수별 당첨 통계 출력
Rank : 당첨 순위 정보 도메인
LottoTest : 당첨순위 확인, 당첨금액 확인 테스트코드 작성

hyunhyun committed:
feat : 당첨로또 입력받기, 유효성 검사 추가

hyunhyun committed:
refactor : 변수 final 설정, 미사용 getter 삭제

hyunhyun committed:
refactor : LottoMachine 삭제

hyunhyun committed:
bug : 로또개수 생성 오류 수정

hyunhyun committed:
test : 수익률 테스트코드

hyunhyun committed:
feat : 수익률 출력

hyunhyun committed:
feat : 로또 메인 함수 생성

jaeyeonling commented on src/main/java/lotto/Lotto.java:
```diff
@@ -0,0 +1,59 @@
+        Collections.sort(lottoNumbers);
+    }
+
+    public void printLotto() {
```
도메인 객체 내부에서 View를 가지고 있다면 어떠한 문제가 있을지 고민해보아요.

hyunhyun replied:
printLotto 에 보면 "[]"과 ", " 등 출력하기 위한 문자열 상수들이 있는데, 출력 양식 변경 위해서 해당 string 변경하려해도 해당 객체 수정하는게 단일 책임 원칙에 어긋나서 일까요??

jaeyeonling commented on src/main/java/lotto/Lotto.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    public Lotto() {
+        createRandomLottoNumber();
```
객체 내부에 랜덤한 요소가 있다면 어떠한 문제가 있을지 고민해보아요.

hyunhyun replied:
객체 단위부터 랜덤한 요소가 있으면 테스트가 어려워서 일까요? 랜덤 생성하는 부분을 더 바깥쪽에서 구현해야될까요?

jaeyeonling replied:
다른 생성자를 통해 테스트 할 수 있기 때문에 선택의 문제긴 하지만, 지금처럼 쉽게 외부로 분리할 수 있다면 분리하는 것을 추천합니다!

jaeyeonling commented on src/main/java/lotto/Lottos.java:
```diff
@@ -0,0 +1,76 @@
+
+public class Lottos {
+
+    final List<Lotto> lottos;
```
`Lottos.lottos`라는 이름이 조금 어색하게 느껴지네요.
중복되지 않는 이름으로 변경해보면 어떨까요?

hyunhyun replied:
lottoList로 변경하겠습니다.

jaeyeonling commented on src/main/java/lotto/Lottos.java:
```diff
@@ -0,0 +1,76 @@
+import java.util.List;
+import java.util.Map;
+
+public class Lottos {
```
해당 객체가 너무 많은 역할을 가지고 있는 것 같네요~
분리할 수 있는 역할이 없을지 고민해보아요.

hyunhyun replied:
당첨개수, 당첨금액 관리하는 Rank의 Map 사용하는 부분들 기능이 서로 비슷해 보여 해당 기능들을 분리하겠습니다.

jaeyeonling commented on src/main/java/lotto/Rank.java:
```diff
@@ -0,0 +1,29 @@
+public class Rank {
+    private final int key;
+    private final int price;
+    int count;
```
접근제어자를 생략한 것은 의도된 내용일까요?

hyunhyun replied:
아래 코멘트 하신 내용과 관련이 있는데, 테스트에서 호출 시 테스트용으로만 사용되는 getter를 사용할지 default 접근제어자로 하여 직접 호출할지 고민하다가 getter로 하였는데 수정이 안된거 같습니다.

jaeyeonling replied:
아직도 수정이 안된 것 같네요~

jaeyeonling commented on src/main/java/lotto/Rank.java:
```diff
@@ -0,0 +1,29 @@
+        count++;
+    }
+
+    public int getCount() {
```
해당 메서드는 테스트를 위한 기능일까요?
테스트를 위한 기능이라면 해당 메서드 없이 테스트할 수 있는 방법이 없을지 고민해보면 좋겠습니다!

hyunhyun replied:
위에서 접근제어자 관련해서 질문을 주셨는데요. 테스트용으로만 사용되는 메소드를 만드는 거보다 테스트용도를 위해서 private으로 할 수 있는 변수를 package-private 변경하는게 더 나은 방법인가요? 
아니면 둘 다 추천하지 않는 방법이고, 테스트를 다른 방식으로 진행해야하는 걸까요?...

jaeyeonling commented on src/test/java/lotto/LottoTest.java:
```diff
@@ -0,0 +1,133 @@
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+public class LottoTest {
```
Lotto에 대한 테스트와 Lottos에 대한 테스트가 섞여있네요!
두 테스트를 분리해보면 좋겠습니다 :)

hyunhyun replied:
입력값 검증 테스트, Lotto 테스트, Lottos 테스트 분리하겠습니다.

jaeyeonling reviewed (changes requested):
로또 자동을 잘 구현해주셨습니다!
고민해볼 수 있는 코멘트를 많이 남겨두었으니 확인 부탁드리고 어려운 내용이 있다면 편하게 남겨주세요 :)

hyunhyun committed:
reafctor :

- test 클래스 분리
- Rank : getCount 메소드 삭제
- WinRank : Lottos 기능 분리
- ResultView : Lotto,Lottos print 분리

hyunhyun committed:
refactor : 오타 수정

jaeyeonling commented on src/main/java/lotto/Lotto.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.Collections;
+import java.util.List;
+
+public class Lotto {
```
Q. printLotto 에 보면 "[]"과 ", " 등 출력하기 위한 문자열 상수들이 있는데, 출력 양식 변경 위해서 해당 string 변경하려해도 해당 객체 수정하는게 단일 책임 원칙에 어긋나서 일까요??
A. 말씀하신 부분도 맞습니다! 추가적으로 다른 View가 추가 됐을 때 어떤 문제가 있을지도 고민해보면 조금 더 명확한 기준이 생기실 것 같습니다 :)

hyunhyun replied:
다른 view 에서도 Lotto의 printLotto를 이용해서 출력을 하다가 각 view 가 출력양식을 바꾸려할 때 결국에 분리하게 되고, 출력하는 부분은 분리되는게 좋은걸까요...?조금 이해가 불명확하게 된거 같아요....

jaeyeonling replied:
View의 형태에 따라 출력되는 양식이 다를 수 있는데요.
출력의 형태에 따라 도메인이 변경되는 것은 안좋은 구조라고 볼 수 있습니다.
추가적으로 ConsoleView는 지금과 같이 내부에서 출력이 가능하지만, WebView라면 지금과 같은 구조로는 출력이 불가능할 것 같아요 ㅎㅎ

jaeyeonling commented on src/main/java/lotto/Lottos.java:
```diff
@@ -0,0 +1,29 @@
+        }
+    }
+
+    public List<Lotto> getLottoList(){
```
```suggestion
    public List<Lotto> getLottoList() {
```

컨벤션을 맞춰주세요~

hyunhyun replied:
넵. 수정하였습니다~

jaeyeonling commented on src/main/java/lotto/Lottos.java:
```diff
@@ -0,0 +1,29 @@
+
+public class Lottos {
+
+    final List<Lotto> lottoList;
```
list와 같은 자료형을 이름으로 사용한다면 어떠한 문제가 있을지 고민해보아요.

hyunhyun replied:
자료형을 다른타입으로 변경하게 되면 변수명도 변경해야하는 문제가 있습니다. 수정하였습니다.

jaeyeonling commented on src/main/java/lotto/Rank.java:
```diff
@@ -0,0 +1,25 @@
+    }
+
+    public void printRank() {
+        System.out.println(key + "개 일치 (" + price + "원)- " + count + "개");
```
https://github.com/next-step/java-lotto-pro/pull/514#discussion_r874401938 해당 코멘트에 대해 고민해보아요.

hyunhyun replied:
출력하는 부분을 ResultView로 분리하였습니다.

jaeyeonling commented on src/main/java/lotto/Rank.java:
```diff
@@ -0,0 +1,25 @@
+package lotto;
+
+public class Rank {
```
Q. 위에서 접근제어자 관련해서 질문을 주셨는데요. 테스트용으로만 사용되는 메소드를 만드는 거보다 테스트용도를 위해서 private으로 할 수 있는 변수를 package-private 변경하는게 더 나은 방법인가요?
아니면 둘 다 추천하지 않는 방법이고, 테스트를 다른 방식으로 진행해야하는 걸까요?...

A. 테스트를 위한 메서드를 만들지 않고 테스트 할 수 있도록 해보면 좋을 것 같아요.
해당 케이스에선 equals를 통해 테스트 해볼 수 있겠네요~

jaeyeonling commented on src/main/java/lotto/WinRanks.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.Map;
+
+public class WinRanks {
+    Map<Integer, Rank> winMap;
```
해당 부분도 접근제어자를 신경써주세요~

hyunhyun replied:
넵 신경쓰도록 하겠습니다~

jaeyeonling commented on src/main/java/lotto/WinRanks.java:
```diff
@@ -0,0 +1,44 @@
+        winMap.put(3, new Rank(3, 5000, 0));
+        winMap.put(4, new Rank(4, 50000, 0));
+        winMap.put(5, new Rank(5, 1500000, 0));
+        winMap.put(6, new Rank(6, 2000000000, 0));
```
```suggestion
        winMap.put(6, new Rank(6, 2_000_000_000, 0));
```

언더바로 구분을 해주면 가독성이 상승합니다~

hyunhyun replied:
넵!!

jaeyeonling commented on src/main/java/lotto/WinRanks.java:
```diff
@@ -0,0 +1,44 @@
+
+    public WinRanks() {
+        winMap = new HashMap<>();
+        winMap.put(3, new Rank(3, 5000, 0));
+        winMap.put(4, new Rank(4, 50000, 0));
+        winMap.put(5, new Rank(5, 1500000, 0));
+        winMap.put(6, new Rank(6, 2000000000, 0));
```
3, 5000
4, 50000 

과 같은 값들은 고정된 값인데요.
고정된 값을 잘 사용하면 해당 객체를 조금 더 가볍게 만들어볼 수 있겠네요~

hyunhyun replied:
순위, 금액을 enum으로 만들었습니다.

jaeyeonling commented on src/main/java/lotto/ui/InputView.java:
```diff
@@ -0,0 +1,67 @@
+        }
+    }
+
+    private static int validateWinningLottoNumber(String lottoNumber) {
```
해당 검증을 View에서 하는 것과 도메인에서 하는 것은 어떠한 차이가 있을지 고민해보아요.

hyunhyun replied:
Lotto와 관련된 검증이긴 하지만 사용자로부터 입력을 받을때만 사용되는 부분이라 입력과 관련 InputView 쪽에 놓았는데, 그래도 Lotto 와 관련된 검증이랑 도메인쪽에 놓는게 이후 다른 부분에서 Lotto를 호출할때 재사용할 수 있어 더 나은 방법일까요?

jaeyeonling commented on src/test/java/lotto/WinRanksTest.java:
```diff
@@ -0,0 +1,83 @@
+        assertThat(winPriceMap.get(5).count).isEqualTo(1);
+        assertThat(winPriceMap.get(6).count).isEqualTo(1);
+
+        ResultView.printLottoResult(lottos, winningLotto);
```
출력에 대한 부분은 어떠한 이유로 추가하셨을까요~?

hyunhyun replied:
테스트하면서 출력되는것을 눈으로 확인하려고 추가하였는데 테스트 통과 확인시에는 불필요하여 삭제하였습니다.

jaeyeonling reviewed (changes requested):
코멘트를 잘 반영해주셨습니다 👍 
추가적인 코멘트 남겨두었으니 확인 부탁드릴게요~

hyunhyun committed:
refactor : 유효성검사 위치 수정

InputView : 로또 유효성 검사 위치 변경으로 호출 수정
Lotto : input 유효성 검사 위치 수정
Lottos : 로또 랜덤생성 RandomUtil 사용으로 수정
LottoTest : 로또 랜덤생성 RandomUtil 사용으로 수정
RandomUtils : 랜덤 로또번호 생성 분리
Rank : print 함수 삭제, 접근제어자 수정
ResultView : rank print 위치 수정
WinRanks : 접근제어자 수정
WinRanksTest : 불필요 출력 삭제

hyunhyun committed:
refactor : Rank enum으로 변경

hyunhyun committed:
refactor : Rank depth 2 해소

jaeyeonling commented on src/test/java/lotto/LottosTest.java:
```diff
@@ -0,0 +1,39 @@
+
+    @BeforeAll
+    public static void createLottosList() {
+        List<Lotto> lottoList = new ArrayList<>();
```
https://github.com/next-step/java-lotto-pro/pull/514#discussion_r875776106 해당 코멘트를 확인해주세요!

jaeyeonling commented on src/test/java/lotto/LottosTest.java:
```diff
@@ -0,0 +1,39 @@
+    @Test
+    public void 금액_입력_로또_개수확인() {
+        Lottos lottos = new Lottos(14000);
+        ResultView.printLottoPurchase(lottos);
```
해당 출력은 디버깅을 위한 것일까요?

jaeyeonling commented on src/main/java/lotto/ui/InputView.java:
```diff
@@ -0,0 +1,52 @@
+        return lotto;
+    }
+
+    private static void validateWinningLottoNumberLength(String[] splitWinningLottoString) {
```
Q. Lotto와 관련된 검증이긴 하지만 사용자로부터 입력을 받을때만 사용되는 부분이라 입력과 관련 InputView 쪽에 놓았는데, 그래도 Lotto 와 관련된 검증이랑 도메인쪽에 놓는게 이후 다른 부분에서 Lotto를 호출할때 재사용할 수 있어 더 나은 방법일까요?

A. 맞습니다. 도메인은 자신의 형태를 보증할 수 있어야 하는데요.
클래스를 설계하고 구현할 때, View에 대한 내용을 고려하지 않고 해당 클래스를 다른 개발자들이 사용할 수 있다고 가정하면서 설계해보시면 좋을 것 같아요.
추가적으로 View에서 검증을 하게 된다면 새로운 View가 추가될 때 마다 View마다 검증을 추가해야 할 수도 있겠네요!

위와 같은 문제가 발생하는 곳을 찾아서 같이 수정해주세요 ㅎㅎ

jaeyeonling commented on src/test/java/lotto/WinRanksTest.java:
```diff
@@ -0,0 +1,80 @@
+
+        WinRanks winRanks = new WinRanks();
+        int winPrice = winRanks.winningPrice(new Lotto(Arrays.asList(1, 2, 3, 4, 5, 6)), lossLottos);
+        double profitRate = ResultView.printProfit(lossLottos.getLottosSize() * 1000, winPrice);
```
해당 역할이 View의 역할이 맞을까요?
View는 단순 출력을 위한 역할로만 사용해보시면 좋을 것 같아요~

jaeyeonling commented on src/main/java/lotto/WinRanks.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.Map;
+
+public class WinRanks {
+    private final Map<Rank, Integer> winMap;
```
https://github.com/next-step/java-lotto-pro/pull/514#discussion_r875776106 해당 코멘트에 대해 고민해보아요.

jaeyeonling reviewed (approved):
코멘트를 잘 반영해주셨습니다 👍 
질문의 답변과 추가적인 코멘트 남겨두었으니 확인 부탁드리고 다음 단계에서 반영해주세요 😄
